ethers-compile/src/compile/compilers/mod.rs

```diff
@@ -0,0 +1,28 @@
+pub mod generic;
+pub mod solc;
+pub mod vyper;
+
+use std::path::PathBuf;
+
+pub use generic::*;
+pub use semver::Version;
+pub use solc::*;
+pub use vyper::*;
+
+use crate::{error::Result, CompilerInput, CompilerOutput};
+use async_trait::async_trait;
+
+#[async_trait]
+pub trait CompilerTrait {
+    fn path(&self) -> PathBuf;
+
+    fn arg(&mut self, arg: String);
+    fn args(&mut self, args: Vec<String>);
+    fn get_args(&self) -> Vec<String>;
+
+    fn version(&self) -> Version;
+    fn language(&self) -> String;
+
+    fn compile_exact(&self, input: &CompilerInput) -> Result<CompilerOutput>;
+    fn compile(&self, input: &CompilerInput) -> Result<CompilerOutput>;
+}
```

Comment:
this is a reasonable first draft.

I don't think we'd need the `arg` related functions.

coming from this comment https://github.com/gakonst/ethers-rs/pull/934#pullrequestreview-888078740 this trait really should do compile only and we want to that be generic over input and output.

so we need something like 

```rust
pub trait Compiler {
  type Input;
  type Output;

  fn compile(&self, input: &Self::Input) -> Result<Self::Output>;
}
```

Reply:
Agree! @joshieDo one way to approach this would be the following:
1. Create the trait mentioned above
2. Impl the trait for ethers-solc
3. Make Foundry's compilation pipeline use the trait
4. Implement the trait for ethers-vyper (or whatever)
5. Create a multi-language implementor of the trait (e.g. instantiated with Ethers-solc, Ethers-vyper, Ethers-fe) which separates sol/vy/fe files, sends them to each job and creates a unified artifact output